### DISCUSSION TITLE
Help getting Apex-Charts right.

### DISCUSSION THREAD

**Discussion by chicagoandy** [BJReplay/ha-solcast-solar#90] [Q&A]
I'm struggling to get my Solar Forecast charting Apex-Charts right.   I started by grabbing the Apex-Charts config from: https://community.home-assistant.io/t/solcast-global-solar-power-forecast-integration/334681/221?u=chicagoandy

Here's a series config:
  - entity: sensor.deye_sunsynk_sol_ark_pv_power
    name: Actual Solar
    float_precision: 3
    color: '#ff9800'
    yaxis_id: kWh
    unit: kW
    transform: return x/1000;
    extend_to: now
    show:
      legend_value: true
      in_header: false
    group_by:
      func: avg
      duration: 5m
 - entity: sensor.solcast_pv_forecast_forecast_today
    name: Solar Forecast (D1)
    extend_to: false
    color: grey
    opacity: 0.3
    stroke_width: 0
    yaxis_id: kWh
    show:
      legend_value: false
      in_header: false
    data_generator: |
      return entity.attributes.detailedForecast.map((entry) => {
            return [new Date(entry.period_start), entry.pv_estimate * 2];
          });

And here's what I get: 
![image](https://github.com/user-attachments/assets/fe0aa458-7616-494e-84b5-d1458d1a9711)

It's clearly wrong.   Todays' forecast is cloudy, D1 forecast should be smaller than yesterdays actual.   I have a 10kw system, so 8kw in the summer is about right.   I have two sites configured in Solcast, one 3.3kw pointing west, and one 7kw pointing south.

On a whim I remove the X2 multiplier, and then the charting was still clearly wrong, values were way too low.

Any suggestions on improving the accuracy of the forecast charts?

**Comment by autoSteve** [comment]
I'm really not sure why the `* 2`, so that's a good 'whim' from my point of view. Remove it, and then compare the  Apex chart to the Home Assistant Energy Dashboard chart. Mine looks _identical_ for a given day.

See an example in the latest readme from @gcoan at https://github.com/BJReplay/ha-solcast-solar.

```
      return entity.attributes.detailedForecast.map((entry) => {
            return [new Date(entry.period_start), entry.pv_estimate];
          });
```

His sample works well for me with some mods for my taste.

Your actuals will not track to the forecast. Actuals are actuals, and forecasts are forecasts (for an average thirty minute period in this case)... They are computed/projected/imagined/maybe until you know what reading actually happened. It's like asking, "did the bureau get the temp right yesterday at 10AM?" Yeah, nah. They were 2C off... But they were pretty close, and I did not need a sweater, which is all I needed to know. Solar forecasting is even less precise given localised cloud action. I see **wild** swings of point-in-time versus average forecast for a half-hour period, which you'd expect when a cloud passes the sun. Unless it's completely sunny that is, then the forecast average from Solcast is pretty darn good.

Always remember that your actuals are point-in-time, but the projections are half-hour averages for that period, smoothed into a spline curve by Apex. In your mind, get rid of the smooth curve shown and replace it with thirty minute average intervals as a 'step' chart. Some minutes in the period may be low, some minutes high, but the overall average for thirty minutes should land at the mid-point on the spline curve for a given thirty minutes.

Assuming Solcast got the average right.

So clearly wrong as "too low"? Probably not when you compare the averages of actual readings.

HTH.

**Comment by chicagoandy** [comment]
There was a forum post which said we needed X2 as the SolCast was only 30 minutes, and we should double it for the hour.   Removing the X2 I get the following: 

      return entity.attributes.detailedForecast.map((entry) => {
            return [new Date(entry.period_start), entry.pv_estimate];
          });

![image](https://github.com/user-attachments/assets/044f7006-200e-417f-8959-5ec78f8b1534)

I typically overproduce on the Energy Dashboard chart as well... 
Here's todays: 
![image](https://github.com/user-attachments/assets/d957a04b-df08-4644-a721-52796b020129)

And here's Yesterdays actuals,  much higher:
![image](https://github.com/user-attachments/assets/cf6057c7-86a5-461f-b022-3b6564101097)

So I guess this is suggesting my problem is in the SolCast config.
South array: 
AC Capacity (inverters)  : 15 kW   (string inverter)
DC Capacity (modules) : 7 kW
Azimuth : -172
Tilt: 40

And in Solcast, the chart that they're showing for that Site,
![image](https://github.com/user-attachments/assets/8b4b9422-a5ac-4eca-be12-2562abca76f6)
That chart is itself showing peaks at 5kw, and that's only 2/3 of my system.

West array:
AC Capacity (inverters) : 15 kW
DC Capacity (modules) : 3.3 kW
Azimuth : 84    *this feels wrong, but it matches that SolCast says 90 = West.
Tilt : 40

And Solcast produces this chart for that array:
![image](https://github.com/user-attachments/assets/78ac7ecf-529c-49dd-9e1e-aaddd2c539a7)

So I would think that my total chart should roughly be adding those two together, and if that happened, I'd have numbers that align pretty closely to my actuals.

![image](https://github.com/user-attachments/assets/08e67483-9c39-4806-b265-75da20156781)

Is anyone else using two Sites ?

**Reply by gcoan** [reply]
I'm using two sites, no problems at all.

in the system debug log you should see retrieval being done for both sites, and in the sensor attributes you can see the site totals

the /config/sites.json file should contain both site configuration

**Comment by autoSteve** [comment]
Many are running two sites. I run two sites. Somewhat more balanced than yours, but that is neither here nor there.

The half-hour intervals are an average production for that half hour. To get a **kWh** forecast for a given hour, add the two half-hour average kW for a given hour and then _divide_ by two, which gets the average of averages as an approximated kWh.

It does interest that your Energy Dashboard is showing a lower splined maximum than the Apex chart. Something is possibly off there, as mine are identical.

With the `* 2` removed it looks pretty good to me...

(The chart works without `* 2` because the period for each interval is one half hour on the chart, spline curved.)

Totalling up your two sites at that point in time indicated results in ~5kW, which is what the Apex shows. Pretty much bang on.

**Reply by chicagoandy** [reply]
Total up the two sites in an hour or so at the peak of the charts gives ~7kW, but Apex-charts is only showing 5.
![image](https://github.com/user-attachments/assets/4312ddd7-7cfd-452c-a3de-40f3c991efa2)   Peak of 4.74

Add 1.97 from this one:
![image](https://github.com/user-attachments/assets/53f537c6-fa4e-43bf-a156-b6dd053662b4)

It should show 6.71.   Maybe not since the times are off.

Argh.   Now I'm going to go model the JSON in excel.

**Comment by chicagoandy** [comment]
![image](https://github.com/user-attachments/assets/ba82d7c4-4d20-4868-837c-4187f489fbef)


Here's the excel - forumulas for Combined fields are simple addition.

https://docs.google.com/spreadsheets/d/e/2PACX-1vSoxDxIZSyekB1Qwk84Z8iX3dTd-KiwuyHil7cw_FSV19ar5l7qwB8rUgL6_jGSmYMnnWuNzn9LtxBF/pubhtml

It should be well over six, right?

If the chart showed six and I overproduced to 7, yeah - that's estimation vs reality.   but the chart is showing 5, and I'm producing close to 8.

<edit>  and charted.

Blue line for Combined:
![image](https://github.com/user-attachments/assets/dcb0cecf-ebce-44e1-8267-1312fed1c380)

**Comment by chicagoandy** [comment]
Jeeze.... Now the chart is refreshed.   I'm going nuts here.   It does Seem to match my excel chart reasonably.

![image](https://github.com/user-attachments/assets/9162f664-e071-4ba1-8758-cc3b301999d2)

So I'm with you the Integration is working properly.   Maybe i"m just coming around to "forecasting is hard", or less generously, doesn't work very well.  

Which brings me back to the original question.... how can I make these charts more accurate?   Just boost the Solcast system size until the charts align?

**Reply by chicagoandy** [reply]
So here's my question.

Why is this chart so far wrong every single day?

![image](https://github.com/user-attachments/assets/26966afc-c36b-436e-b43b-429f01367b91)

I acknowledge these questions are probably better directed at Solcast.   I am curious if anyone else has seen similar variance.   I do see this variance daily, and it is a rare day with un-forcasted weather that I don't overperform substantially.

**Comment by autoSteve** [comment]
~~There is something funky going on with writing solcast-sites.json, @chicagoandy, so I don't think you're slowly going mad.~~

~~I just checked mine and it only contains one site for some reason.~~

~~If the integration relied on the data in this file on startup (like if it hit a 429/too busy) then the forecast would definitely be out, which would account for the difference. It checks in with Solcast at every start, so only temporary failures would impact operation, but that situation would not mend itself until restart.~~

Check the contents of this file in your case.

I'm off to stare at some code.

**Reply by gcoan** [reply]
My solcast-sites.json still contains 2 sites (running alpha 2):

```
{"sites": [{"name": "87 High Street (E)", "resource_id": "bc31-4d4c-xxxx-xxxx", "capacity": 10, "capacity_dc": 8.58, "longitude": -xxxx, "latitude": xxxx, "azimuth": -100, "tilt": 37, "install_date": "2023-01-07T00:00:00.0000000Z", "loss_factor": 0.99, "tags": ["UK", "Bedfordshire", "GivEnergy"]}, {"name": "87 High Street (W)", "resource_id": "23bb-xxxx-xxxx-fd94", "capacity": 9, "capacity_dc": 6.34, "longitude": -xxxx, "latitude": xxxx, "azimuth": 80, "tilt": 37, "install_date": "2023-01-07T00:00:00.0000000Z", "loss_factor": 0.98, "tags": ["UK", " Bedfordshire", " GivEnergy"]}], "page_count": 1, "current_page": 1, "total_records": 2}
```

**Reply by autoSteve** [reply]
Actually no. Am I going slowly mad?

Re-checked and it looks fine. Both sites are in there, and the file written date has not changed. 🤷‍♂️

**Reply by autoSteve** [reply]
It's the bloody mRemoteNG SSH terminal that threw me off... A bug in that piece of cr*p program when running in Parallels.

What I did: Open terminal, cat the file, only see one site. Look at code and conclude this is impossible. Slightly change how big the mRemoteNG window is, re-cat the file and I see it all. 🤦🏻

**Reply by gcoan** [reply]
> It's the bloody mRemoteNG SSH terminal that threw me off... A bug in that piece of cr*p program when running in Parallels.
> 
> 
> 
> What I did: Open terminal, cat the file, only see one site. Look at code and conclude this is impossible. Slightly change how big the mRemoteNG window is, re-cat the file and I see it all. 🤦🏻

face plant indeed

I never SSH to my HA box.  Either access it as a network drive share or use studio code server from HA to edit files directly. 
Don't know why I still run the SSH add-on really...

**Comment by autoSteve** [comment]
Verify in the HA logs that your two sites are actually being acquired when the update automation runs,  #@chicagoandy. This occurs sequentially, so you will see one set of log lines for each site.

Or (if you're not running v4.0.39-alpha and have turned off the site breakdown), there should be separate attributes for each site in many of the sensors giving the breakdown, named after the site ID. If only one site is being acquired then you'll only see one ID.

**Reply by chicagoandy** [reply]
Confirming both sites are in solcast-sites.json.
{"sites": [{"name": "Home", "resource_id": "XXX-12a4", "capacity": 15, "capacity_dc": 7, "longitude": -84.XX, "latitude": 33.XX, "azimuth": -172, "tilt": 40, "install_date": "2024-06-07T04:00:00.0000000Z", "loss_factor": 1, "tags": []}, {"name": "west", "resource_id": "XXX-8c52", "capacity": 15, "capacity_dc": 3.3, "longitude": -84.XXX, "latitude": 33.XXX, "azimuth": 84, "tilt": 40, "install_date": "2024-06-07T04:00:00.0000000Z", "loss_factor": 1, "tags": []}], "page_count": 1, "current_page": 1, "total_records": 2}#

And yes, seeing both in the logfile.    Thank you for your effort, I was hopeful I could find something actually wrong, but apparently not...

For giggles, (maybe sympathy), here's todays charts:
![image](https://github.com/user-attachments/assets/267c71ee-56d4-4a04-b3f4-da2b02ec01f0)

**Comment by autoSteve** [comment]
In looking at this again I am still confused by how inacurate the forecast is consistently being for you.

What's your set-up, by the way? You have two sites, so two separate strings of panels facing different directions, and a single inverter?

I have not been able to confirm how Solcast use the DC & AC capacity values. I have micro-inverters, so setting up Solcast is super simple. For one site I have 7.92 kW DC capacity, and 6.857 kW AC capacity. The other is 5.72 kW DC and 5.243 kW AC. Both set from the spec sheets of the panels/inverters, and in both cases a lower AC output than DC.

You have a much higher AC than DC capacity, and I'm wondering whether this is messing with forecasts.

Just a guess, because as I said I can't find documentation regarding exactly how these settings influence the Solcast projections.

**Reply by gcoan** [reply]
I have 6 strings on two orientations and 3 inverters.  Two inverters are hybrid inverters, one is an AC inverter. These are setup as two sites in Solcast, both of which have AC capacity higher than the DC capacity so there shouldn't be any Solcast clipping (in practice this config is not quite right as one inverter has its two strings on opposite sides of the roof but I duplicate the AC capacity in the two sites,  but in HA I don't see any evidence of inverter clipping so I think its ok).

But consistently I very rarely achieve my Solcast PV forecast, its usually somewhere around 40-50% between the PV10 and PV50 figures.

e.g. first three weeks of July:
![image](https://github.com/user-attachments/assets/e028f33a-fa4e-42da-865f-6df4f3d5fe50)

I've just got used to that's the way it is.  Some of the summer under performance I know is due to high solar export pushing the grid voltage over limit and two of the inverters stop generation briefly when this happens.  Some of the under performance I suspect is temperature related, hot panels in summer are less efficient.  In winter I see solar generation exceeding the forecast.  And the rest of the difference, I have no idea.

**Reply by chicagoandy** [reply]
Thanks for looking into this.

II went and looked back at this and I did notice that I had my efficiency factor set at the default of 90.   Since my system is new, I have now updated this to 100%.   

Todays chart has a much higher estimate.   The change is far greater than a 10% increase in efficiency.   I think they may be miscalculating this adjustment.

![image](https://github.com/user-attachments/assets/0d63d5bb-8318-4f60-b363-5442749c5e54)

Today had non-forecasted cloudiness in the afternoon.    So I am still producing over the estimates in the morning, and I am producing under the estimates this afternoon (due to cloudiness).   So I'm on-track for today, despite the afternoon cloudiness.

Yes, I have two strings, one facing West, the other facing South.

18 X REC 420 panels, on the roof facing South (actually 172).   Estimating 40 degree tilt.   18 X 420 = 7,560 watts
8 X REC 420 panels, on the roof facing West (actually +84)  Same 40 degree tilt.    8 X 420 = 3,360 watts

A single Sol-Ark 15K inverter.   3 MPPTs, each string is under the MPPT capacity.    Sol-Ark can actually handle 19Kw of PV.    Clipping is never an issue.

I have these defined in Sol-cast as:

South:
![image](https://github.com/user-attachments/assets/ab34a9a6-c795-4538-ac02-29de16d19fd2)

West:
![image](https://github.com/user-attachments/assets/83bd59f9-71b9-49e6-a0b9-dd897202e8f0)

**Reply by gcoan** [reply]
The solcast configuration looks correct.

There's always going to be some variation in the generation each day. Might be worth you looking at the PV10 and PV50 figures in HA to get an idea of the range of forecast variability.

**Reply by autoSteve** [reply]
Azimuth and tilt set correctly @chicagoandy?
